### PR TITLE
Support tuple metaclass indexers with non-literal arguments

### PR DESCRIPTION
**Pull request by HertzDevil**
Implements the following class methods:

* `Tuple.[](index : Int)`
* `Tuple.[]?(index : Int)`
* `NamedTuple.[](key : Symbol | String)`
* `NamedTuple.[]?(key : Symbol | String)`

They already work when the index is an integer literal, or when the key is a symbol or string literal. With this PR regular, non-literal variables also work.

Also "adds" a `Tuple.[](range : Range)` overload which always fails to compile unless `range` is a literal, similar to its non-metaclass counterpart.

### CODE REVIEW

File: src/named_tuple.cr [hunk 1]
```diff
@@ -124,24 +154,94 @@ struct NamedTuple
...
-  # tuple[key]? # => "Crystal"
+  # tuple[key]?         # => "Crystal"
+  # typeof(tuple[key]?) # => (Int32 | String | Nil)
   #
   # key = "year"
   # tuple[key] # => 2011
   #
   # key = :other
-  # tuple[key]? # => nil
+  # tuple[key] # => nil
```
Comment (Sija) [inline src/named_tuple.cr:204]: That looks like a mistake.

File: src/named_tuple.cr [hunk 2]
```diff
@@ -124,24 +154,94 @@ struct NamedTuple
...
+  def self.[]?(key : Symbol | String)
+    {% begin %}
+      case key
+      {% for key in T %}
+      when {{ key.symbolize }}, {{ key.stringify }}
+        typeof(begin
+          x = uninitialized self
+          x[{{ key.symbolize }}]
+        end)
+      {% end %}
```
Comment (asterite) [inline src/named_tuple.cr:240]:
```suggestion
      {% for key, value in T %}
      when {{ key.symbolize }}, {{ key.stringify }}
        {{value}}
      {% end %}
```

File: src/tuple.cr [hunk 3]
```diff
@@ -175,23 +196,87 @@ struct Tuple
...
+  # ```
+  def self.[]?(index : Int)
+    {% begin %}
+      case index
+      {% for i in 0...T.size %}
+      when {{ i }}, {{ i - T.size }}
+        typeof(begin
+          v = uninitialized self
+          v[{{ i }}]
+        end)
```
[the remark this answers is not in the document]
Comment (HertzDevil) [reply src/tuple.cr:268]: This form allows (file-)private types to be used. It is the same workaround as `Tuple.new`. Ditto for `NamedTuple`.

[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/tuple.cr:268]: Can we add a comment about that?

**Comment by beta-ziliani** [comment]
Thinking of #12005, to me the pattern

```crystal
FIELD = :field
tuple[FIELD]
```

makes perfect sense; however, it won't work as expected since it will return the union of all the types of the tuple even if the value is known at compile time. How hard is it to consider this case specially?

**Comment by asterite** [comment]
> How hard is it to consider this case specially?

Very easy! But is there a reason to discuss this here instead of in #12005?

**Comment by beta-ziliani** [comment]
No, it doesn't make sense for it to be here. (me to brain rn: GH comments are not a chat system GH comments are not a chat system GH comments are not a chat system ....)

File: src/named_tuple.cr [hunk 4]
```diff
@@ -122,16 +153,51 @@ struct NamedTuple
...
+  # alias Foo = NamedTuple(name: String, year: Int32)
+  #
+  # Foo[:name]       # => String
+  # Foo["year"]      # => Int32
+  # Foo["year"].zero # => 0
+  # Foo[:other]      # => Error: missing key 'other' for named tuple NamedTuple(name: String, year: Int32).class
+  #
+  # key = :year
+  # Foo[key]      # => Int32
+  # Foo[key].zero # Error: undefined method '[]' for NamedTuple(name: String, year: Int32).class
```
Comment (beta-ziliani) [inline src/named_tuple.cr:171]: Shouldn't it be:
```suggestion
  # Foo[key].zero # Error: undefined method 'zero' for String.class (compile-time type is (Int32.class | String.class | Nil))
```

**Review by beta-ziliani — CHANGES_REQUESTED** [review]
Thanks! Just one very minor concern.

File: src/named_tuple.cr (hunk 4 above)
Comment (HertzDevil) [reply src/named_tuple.cr]: it's not nilable (`self.[]?` is), other than that yes

Comment (beta-ziliani) [reply src/named_tuple.cr]: Wrong copy-paste, sorry!

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**